File: packages/parser/src/parser.ts
import type {
  ArrayExpr,
  Diagnostic,
  Expr,
  InputDecl,
  MappingFile,
  ObjectExpr,
  ObjectProp,
  Span,
  TypeFieldNode,
  TypeNode,
} from "@thomasfarineau/typeflow-core";
import { LexError, tokenize, type Token } from "./lexer.ts";

export interface ParseResult {
  ast: MappingFile | null;
  diagnostics: Diagnostic[];
}

class ParseError extends Error {
  constructor(public diagnostic: Diagnostic) {
    super(diagnostic.message);
  }
}

const PRIM_NAMES = new Set(["string", "number", "boolean", "null", "unknown"]);

export function parse(source: string): ParseResult {
  let tokens: Token[];
  try {
    tokens = tokenize(source);
  } catch (e) {
    if (e instanceof LexError) return { ast: null, diagnostics: [e.diagnostic] };
    throw e;
  }
  try {
    const ast = new Parser(tokens).parseFile();
    return { ast, diagnostics: [] };
  } catch (e) {
    if (e instanceof ParseError) return { ast: null, diagnostics: [e.diagnostic] };
    throw e;
  }
}

class Parser {
  private pos = 0;

  constructor(private tokens: Token[]) {}

  private peek(offset = 0): Token {
    return this.tokens[Math.min(this.pos + offset, this.tokens.length - 1)]!;
  }

  private next(): Token {
    const t = this.peek();
    if (t.type !== "eof") this.pos++;
    return t;
  }

  private at(type: "punct" | "ident", value: string): boolean {
    const t = this.peek();
    return t.type === type && t.value === value;
  }

  private eat(type: "punct" | "ident", value: string): boolean {
    if (this.at(type, value)) {
      this.pos++;
      return true;
    }
    return false;
  }

  private expectPunct(value: string): Token {
    if (!this.at("punct", value)) this.fail(`Expected '${value}'.`);
    return this.next();
  }

  private expectIdent(): Token {
    const t = this.peek();
    if (t.type !== "ident") this.fail("Expected an identifier.");
    return this.next();
  }

  private fail(message: string, span?: Span): never {
    const t = this.peek();
    throw new ParseError({
      code: "TF1002",
      message: t.type === "eof" ? `${message} Unexpected end of file.` : message,
      span: span ?? { start: t.start, end: t.end },
      severity: "error",
    });
  }

  parseFile(): MappingFile {
    let input: InputDecl | undefined;
    while (this.at("ident", "input")) {
      if (input) this.fail("Only one 'input' declaration is allowed in v0.1.");
      input = this.parseInputDecl();
    }
    if (!this.at("ident", "map")) {
      this.fail("Expected a 'map { ... }' block.");
    }
    this.next(); // map
    const map = this.parseObjectExpr();
    if (this.peek().type !== "eof") {
      this.fail("Unexpected content after the 'map' block.");
    }
    return { input, map };
  }

  private parseInputDecl(): InputDecl {
    const kw = this.next(); // input
    const nameTok = this.expectIdent();
    this.expectPunct(":");

    // `input user: User from "./types"` — a bare capitalized identifier followed by `from`
    // is a type reference resolved by a schema adapter.
    if (this.peek().type === "ident" && !PRIM_NAMES.has(this.peek().value) && this.peek(1).type === "ident" && this.peek(1).value === "from") {
      const typeName = this.next().value;
      this.next(); // from
      const fromTok = this.peek();
      if (fromTok.type !== "string") this.fail("Expected a module path string after 'from'.");
      this.next();
      return {
        name: nameTok.value,
        typeRef: { typeName, from: fromTok.value },
        span: { start: kw.start, end: fromTok.end },
      };
    }

    const inlineType = this.parseTypeExpr();
    return {
      name: nameTok.value,
      inlineType,
      span: { start: kw.start, end: inlineType.span.end },
    };
  }

  // ---- Inline type syntax ----

  private parseTypeExpr(): TypeNode {
    const first = this.parsePostfixType();
    if (!this.at("punct", "|")) return first;
    const types = [first];
    while (this.eat("punct", "|")) types.push(this.parsePostfixType());
    return { kind: "union", types, span: { start: first.span.start, end: types[types.length - 1]!.span.end } };
  }

  private parsePostfixType(): TypeNode {
    let t = this.parsePrimaryType();
    while (this.at("punct", "[") && this.peek(1).type === "punct" && this.peek(1).value === "]") {
      this.next();
      const close = this.next();
      t = { kind: "array", element: t, span: { start: t.span.start, end: close.end } };
    }
    return t;
  }

  private parsePrimaryType(): TypeNode {
    const t = this.peek();
    if (t.type === "ident") {
      if (PRIM_NAMES.has(t.value)) {
        this.next();
        return { kind: "prim", name: t.value as never, span: { start: t.start, end: t.end } };
      }
      if (t.value === "true" || t.value === "false") {
        this.next();
        return { kind: "lit", value: t.value === "true", span: { start: t.start, end: t.end } };
      }
      this.fail(`Unknown type '${t.value}'. Inline types support: string, number, boolean, null, unknown, literals, arrays, objects, and unions.`);
    }
    if (t.type === "string") {
      this.next();
      return { kind: "lit", value: t.value, span: { start: t.start, end: t.end } };
    }
    if (t.type === "number") {
      this.next();
      return { kind: "lit", value: Number(t.value), span: { start: t.start, end: t.end } };
    }
    if (this.at("punct", "(")) {
      this.next();
      const inner = this.parseTypeExpr();
      this.expectPunct(")");
      return inner;
    }
    if (this.at("punct", "{")) {
      const open = this.next();
      const fields: TypeFieldNode[] = [];
      while (!this.at("punct", "}")) {
        const nameTok = this.peek();
        let name: string;
        if (nameTok.type === "ident") name = this.next().value;
        else if (nameTok.type === "string") name = this.next().value;
        else this.fail("Expected a field name.");
        const optional = this.eat("punct", "?");
        this.expectPunct(":");
        const type = this.parseTypeExpr();
        fields.push({ name: name!, optional, type, span: { start: nameTok.start, end: type.span.end } });
        if (!this.eat("punct", ",")) break;
      }
      const close = this.expectPunct("}");
      return { kind: "object", fields, span: { start: open.start, end: close.end } };
    }
    this.fail("Expected a type.");
  }

  // ---- Expressions ----

  private parseObjectExpr(): ObjectExpr {
    const open = this.expectPunct("{");
    const props: ObjectProp[] = [];
    while (!this.at("punct", "}")) {
      const nameTok = this.peek();
      let name: string;
      if (nameTok.type === "ident") name = this.next().value;
      else if (nameTok.type === "string") name = this.next().value;
      else this.fail("Expected a property name.");
      this.expectPunct(":");
      const value = this.parseExpr();
      props.push({ name: name!, value, span: { start: nameTok.start, end: value.span.end } });
      if (!this.eat("punct", ",")) break;
    }
    const close = this.expectPunct("}");
    return { kind: "object", props, span: { start: open.start, end: close.end } };
  }

  parseExpr(): Expr {
    return this.parseTernary();
  }

  private parseTernary(): Expr {
    const cond = this.parseNullish();
    if (!this.eat("punct", "?")) return cond;
    const then = this.parseExpr();
    this.expectPunct(":");
    const els = this.parseExpr();
    return { kind: "cond", cond, then, else: els, span: { start: cond.span.start, end: els.span.end } };
  }

  private parseBinaryLevel(ops: string[], nextLevel: () => Expr): Expr {
    let left = nextLevel.call(this);
    while (this.peek().type === "punct" && ops.includes(this.peek().value)) {
      const op = this.next().value as never;
      const right = nextLevel.call(this);
      left = { kind: "binary", op, left, right, span: { start: left.span.start, end: right.span.end } };
    }
    return left;
  }

  private parseNullish(): Expr {
    return this.parseBinaryLevel(["??"], this.parseOr);
  }

  private parseOr(): Expr {
    return this.parseBinaryLevel(["||"], this.parseAnd);
  }

  private parseAnd(): Expr {
    return this.parseBinaryLevel(["&&"], this.parseEquality);
  }

  private parseEquality(): Expr {
    return this.parseBinaryLevel(["==", "!="], this.parseRelational);
  }

  private parseRelational(): Expr {
    return this.parseBinaryLevel(["<", "<=", ">", ">="], this.parseAdditive);
  }

  private parseAdditive(): Expr {
    return this.parseBinaryLevel(["+", "-"], this.parseMultiplicative);
  }

  private parseMultiplicative(): Expr {
    return this.parseBinaryLevel(["*", "/"], this.parseUnary);
  }

  private parseUnary(): Expr {
    const t = this.peek();
    if (this.at("punct", "!") || this.at("punct", "-")) {
      this.next();
      const operand = this.parseUnary();
      return { kind: "unary", op: t.value as "!" | "-", operand, span: { start: t.start, end: operand.span.end } };
    }
    return this.parsePostfix();
  }

  private parsePostfix(): Expr {
    let expr = this.parsePrimary();
    for (;;) {
      if (this.at("punct", ".") || this.at("punct", "?.")) {
        const optional = this.peek().value === "?.";
        this.next();
        const nameTok = this.expectIdent();
        expr = {
          kind: "member",
          object: expr,
          name: nameTok.value,
          optional,
          span: { start: expr.span.start, end: nameTok.end },
          nameSpan: { start: nameTok.start, end: nameTok.end },
        };
      } else if (this.at("punct", "[")) {
        this.next();
        const inner = this.parseExpr();
        const close = this.expectPunct("]");
        expr = { kind: "bracket", object: expr, inner, span: { start: expr.span.start, end: close.end } };
      } else if (this.at("punct", "->")) {
        this.next();
        const body = this.parseObjectExpr();
        expr = { kind: "project", object: expr, body, span: { start: expr.span.start, end: body.span.end } };
      } else {
        return expr;
      }
    }
  }

  private parsePrimary(): Expr {
    const t = this.peek();

    if (t.type === "number") {
      this.next();
      return { kind: "lit", value: Number(t.value), span: { start: t.start, end: t.end } };
    }
    if (t.type === "string") {
      this.next();
      return { kind: "lit", value: t.value, span: { start: t.start, end: t.end } };
    }
    if (t.type === "ident") {
      if (t.value === "true" || t.value === "false") {
        this.next();
        return { kind: "lit", value: t.value === "true", span: { start: t.start, end: t.end } };
      }
      if (t.value === "null") {
        this.next();
        return { kind: "lit", value: null, span: { start: t.start, end: t.end } };
      }
      this.next();
      // Function call: `name(args)`
      if (this.at("punct", "(")) {
        this.next();
        const args: Expr[] = [];
        while (!this.at("punct", ")")) {
          args.push(this.parseExpr());
          if (!this.eat("punct", ",")) break;
        }
        const close = this.expectPunct(")");
        return {
          kind: "call",
          name: t.value,
          args,
          span: { start: t.start, end: close.end },
          nameSpan: { start: t.start, end: t.end },
        };
      }
      return { kind: "ident", name: t.value, span: { start: t.start, end: t.end } };
    }
    if (this.at("punct", "(")) {
      this.next();
      const inner = this.parseExpr();
      const close = this.expectPunct(")");
      inner.span = { start: t.start, end: close.end };
      return inner;
    }
    if (this.at("punct", "{")) {
      return this.parseObjectExpr();
    }
    if (this.at("punct", "[")) {
      this.next();
      const elements: Expr[] = [];
      while (!this.at("punct", "]")) {
        elements.push(this.parseExpr());
        if (!this.eat("punct", ",")) break;
      }
      const close = this.expectPunct("]");
      return { kind: "array", elements, span: { start: t.start, end: close.end } } as ArrayExpr;
    }
    this.fail("Expected an expression.");
  }
}

File: apps/converters/src/tf/parser.rs
//! Typeflow parser (port of `src/parser/parser.ts`) — the subset needed by
//! the formatter: full file parsing, no checker.

use super::ast::*;
use super::lexer::{tokenize, LexError, Token, TokenType, TokenizeOptions};

pub struct ParseError {
    pub message: String,
}

pub struct ParseOutcome {
    pub ast: Option<MappingFile>,
    /// First diagnostic message when parsing failed.
    pub error: Option<String>,
}

const PRIM_NAMES: &[&str] = &["string", "number", "boolean", "null", "unknown"];

pub fn parse(source: &str) -> ParseOutcome {
    let tokens = match tokenize(
        source,
        &TokenizeOptions {
            include_comments: false,
            tolerant: false,
        },
    ) {
        Ok(t) => t,
        Err(LexError { message }) => {
            return ParseOutcome {
                ast: None,
                error: Some(message),
            }
        }
    };
    let mut parser = Parser { tokens, pos: 0 };
    match parser.parse_file() {
        Ok(ast) => ParseOutcome {
            ast: Some(ast),
            error: None,
        },
        Err(ParseError { message }) => ParseOutcome {
            ast: None,
            error: Some(message),
        },
    }
}

struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

type PResult<T> = Result<T, ParseError>;

impl Parser {
    fn peek(&self) -> &Token {
        self.peek_at(0)
    }

    fn peek_at(&self, offset: usize) -> &Token {
        let idx = std::cmp::min(self.pos + offset, self.tokens.len() - 1);
        &self.tokens[idx]
    }

    fn next(&mut self) -> Token {
        let t = self.peek().clone();
        if t.type_ != TokenType::Eof {
            self.pos += 1;
        }
        t
    }

    fn at(&self, type_: TokenType, value: &str) -> bool {
        let t = self.peek();
        t.type_ == type_ && t.value == value
    }

    fn eat(&mut self, type_: TokenType, value: &str) -> bool {
        if self.at(type_, value) {
            self.pos += 1;
            true
        } else {
            false
        }
    }

    fn expect_punct(&mut self, value: &str) -> PResult<Token> {
        if !self.at(TokenType::Punct, value) {
            return Err(self.fail(&format!("Expected '{}'.", value)));
        }
        Ok(self.next())
    }

    fn expect_ident(&mut self) -> PResult<Token> {
        if self.peek().type_ != TokenType::Ident {
            return Err(self.fail("Expected an identifier."));
        }
        Ok(self.next())
    }

    fn fail(&self, message: &str) -> ParseError {
        let t = self.peek();
        ParseError {
            message: if t.type_ == TokenType::Eof {
                format!("{} Unexpected end of file.", message)
            } else {
                message.to_string()
            },
        }
    }

    fn parse_file(&mut self) -> PResult<MappingFile> {
        let mut input: Option<InputDecl> = None;
        let mut uses: Vec<UseDecl> = Vec::new();
        let mut fns: Vec<FnDecl> = Vec::new();
        loop {
            if self.at(TokenType::Ident, "input") {
                if input.is_some() {
                    return Err(self.fail("Only one 'input' declaration is allowed in v0.1."));
                }
                input = Some(self.parse_input_decl()?);
            } else if self.at(TokenType::Ident, "use") {
                uses.push(self.parse_use_decl()?);
            } else if self.at(TokenType::Ident, "fn") {
                fns.push(self.parse_fn_decl()?);
            } else {
                break;
            }
        }
        if !self.at(TokenType::Ident, "map") {
            return Err(self.fail("Expected a 'map { ... }' block."));
        }
        self.next(); // map
        let map = self.parse_object_expr()?;
        if self.peek().type_ != TokenType::Eof {
            return Err(self.fail("Unexpected content after the 'map' block."));
        }
        Ok(MappingFile {
            input,
            uses,
            fns,
            map,
        })
    }

    fn parse_param_list(&mut self) -> PResult<Vec<UseParam>> {
        self.expect_punct("(")?;
        let mut params = Vec::new();
        while !self.at(TokenType::Punct, ")") {
            let p_tok = self.expect_ident()?;
            let optional = self.eat(TokenType::Punct, "?");
            self.expect_punct(":")?;
            let type_ = self.parse_type_expr()?;
            params.push(UseParam {
                name: p_tok.value,
                optional,
                type_,
            });
            if !self.eat(TokenType::Punct, ",") {
                break;
            }
        }
        self.expect_punct(")")?;
        Ok(params)
    }

    fn parse_signature(&mut self) -> PResult<(String, Vec<UseParam>, TypeNode, usize)> {
        let name_tok = self.expect_ident()?;
        let params = self.parse_param_list()?;
        self.expect_punct(":")?;
        let return_type = self.parse_type_expr()?;
        Ok((name_tok.value, params, return_type, name_tok.start))
    }

    fn parse_fn_decl(&mut self) -> PResult<FnDecl> {
        let kw = self.next(); // fn
        let name_tok = self.expect_ident()?;
        let params = self.parse_param_list()?;
        let mut return_type: Option<TypeNode> = None;
        if self.eat(TokenType::Punct, ":") {
            return_type = Some(self.parse_type_expr()?);
        }
        if !self.at(TokenType::Punct, "=") {
            return Err(self.fail("Expected '=' followed by the function body expression."));
        }
        self.next(); // =
        let body = self.parse_expr()?;
        let span = Span {
            start: kw.start,
            end: body.span().end,
        };
        Ok(FnDecl {
            name: name_tok.value,
            params,
            return_type,
            body,
            span,
        })
    }

    fn parse_use_decl(&mut self) -> PResult<UseDecl> {
        let kw = self.next(); // use
        let (name, params, return_type, _) = self.parse_signature()?;
        if !self.at(TokenType::Ident, "from") {
            return Err(self.fail("Expected 'from \"./module\"' after the function signature."));
        }
        self.next(); // from
        if self.peek().type_ != TokenType::Str {
            return Err(self.fail("Expected a module path string after 'from'."));
        }
        let from_tok = self.next();
        Ok(UseDecl {
            name,
            params,
            return_type,
            from: from_tok.value,
            span: Span {
                start: kw.start,
                end: from_tok.end,
            },
        })
    }

    fn parse_input_decl(&mut self) -> PResult<InputDecl> {
        let kw = self.next(); // input
        let name_tok = self.expect_ident()?;
        self.expect_punct(":")?;

        // `input user: User from "./types"` — a bare non-primitive identifier
        // followed by `from` is a type reference.
        if self.peek().type_ == TokenType::Ident
            && !PRIM_NAMES.contains(&self.peek().value.as_str())
            && self.peek_at(1).type_ == TokenType::Ident
            && self.peek_at(1).value == "from"
        {
            let type_name = self.next().value;
            self.next(); // from
            if self.peek().type_ != TokenType::Str {
                return Err(self.fail("Expected a module path string after 'from'."));
            }
            let from_tok = self.next();
            return Ok(InputDecl {
                name: name_tok.value,
                type_ref: Some((type_name, from_tok.value)),
                inline_type: None,
                span: Span {
                    start: kw.start,
                    end: from_tok.end,
                },
            });
        }

        let inline_type = self.parse_type_expr()?;
        let span = Span {
            start: kw.start,
            end: inline_type.span().end,
        };
        Ok(InputDecl {
            name: name_tok.value,
            type_ref: None,
            inline_type: Some(inline_type),
            span,
        })
    }

    // ---- Inline type syntax ----

    fn parse_type_expr(&mut self) -> PResult<TypeNode> {
        let first = self.parse_postfix_type()?;
        if !self.at(TokenType::Punct, "|") {
            return Ok(first);
        }
        let mut types = vec![first];
        while self.eat(TokenType::Punct, "|") {
            types.push(self.parse_postfix_type()?);
        }
        let span = Span {
            start: types[0].span().start,
            end: types[types.len() - 1].span().end,
        };
        Ok(TypeNode::Union { types, span })
    }

    fn parse_postfix_type(&mut self) -> PResult<TypeNode> {
        let mut t = self.parse_primary_type()?;
        while self.at(TokenType::Punct, "[")
            && self.peek_at(1).type_ == TokenType::Punct
            && self.peek_at(1).value == "]"
        {
            self.next();
            let close = self.next();
            let span = Span {
                start: t.span().start,
                end: close.end,
            };
            t = TypeNode::Array {
                element: Box::new(t),
                span,
            };
        }
        Ok(t)
    }

    fn parse_primary_type(&mut self) -> PResult<TypeNode> {
        let t = self.peek().clone();
        if t.type_ == TokenType::Ident {
            if PRIM_NAMES.contains(&t.value.as_str()) {
                self.next();
                return Ok(TypeNode::Prim {
                    name: t.value,
                    span: Span {
                        start: t.start,
                        end: t.end,
                    },
                });
            }
            if t.value == "true" || t.value == "false" {
                self.next();
                return Ok(TypeNode::Lit {
                    value: LitValue::Bool(t.value == "true"),
                    span: Span {
                        start: t.start,
                        end: t.end,
                    },
                });
            }
            return Err(self.fail(&format!(
                "Unknown type '{}'. Inline types support: string, number, boolean, null, unknown, literals, arrays, objects, and unions.",
                t.value
            )));
        }
        if t.type_ == TokenType::Str {
            self.next();
            return Ok(TypeNode::Lit {
                value: LitValue::Str(t.value),
                span: Span {
                    start: t.start,
                    end: t.end,
                },
            });
        }
        if t.type_ == TokenType::Number {
            self.next();
            return Ok(TypeNode::Lit {
                value: LitValue::Num(t.value.parse::<f64>().unwrap_or(f64::NAN)),
                span: Span {
                    start: t.start,
                    end: t.end,
                },
            });
        }
        if self.at(TokenType::Punct, "(") {
            self.next();
            let inner = self.parse_type_expr()?;
            self.expect_punct(")")?;
            return Ok(inner);
        }
        if self.at(TokenType::Punct, "{") {
            let open = self.next();
            let mut fields: Vec<TypeFieldNode> = Vec::new();
            while !self.at(TokenType::Punct, "}") {
                let name_tok = self.peek().clone();
                let name = if name_tok.type_ == TokenType::Ident || name_tok.type_ == TokenType::Str
                {
                    self.next().value
                } else {
                    return Err(self.fail("Expected a field name."));
                };
                let optional = self.eat(TokenType::Punct, "?");
                self.expect_punct(":")?;
                let type_ = self.parse_type_expr()?;
                let span = Span {
                    start: name_tok.start,
                    end: type_.span().end,
                };
                fields.push(TypeFieldNode {
                    name,
                    optional,
                    type_,
                    span,
                });
                if !self.eat(TokenType::Punct, ",") {
                    break;
                }
            }
            let close = self.expect_punct("}")?;
            return Ok(TypeNode::Object {
                fields,
                span: Span {
                    start: open.start,
                    end: close.end,
                },
            });
        }
        Err(self.fail("Expected a type."))
    }

    // ---- Expressions ----

    fn parse_object_expr(&mut self) -> PResult<ObjectExpr> {
        let open = self.expect_punct("{")?;
        let mut props: Vec<ObjectProp> = Vec::new();
        let mut lets: Vec<LetBinding> = Vec::new();
        while !self.at(TokenType::Punct, "}") {
            // `let name = expr` — keyword form only; `let: ...` stays a property.
            if self.peek().type_ == TokenType::Ident
                && self.peek().value == "let"
                && self.peek_at(1).type_ == TokenType::Ident
                && self.peek_at(1).value != "let"
                && self.peek_at(2).type_ == TokenType::Punct
                && self.peek_at(2).value == "="
            {
                let kw = self.next();
                let name_tok = self.next();
                self.expect_punct("=")?;
                let value = self.parse_expr()?;
                let span = Span {
                    start: kw.start,
                    end: value.span().end,
                };
                lets.push(LetBinding {
                    name: name_tok.value,
                    value,
                    span,
                });
                if !self.eat(TokenType::Punct, ",") {
                    break;
                }
                continue;
            }

            let name_tok = self.peek().clone();
            let name = if name_tok.type_ == TokenType::Ident || name_tok.type_ == TokenType::Str {
                self.next().value
            } else {
                return Err(self.fail("Expected a property name."));
            };
            self.expect_punct(":")?;
            let value = self.parse_expr()?;
            let span = Span {
                start: name_tok.start,
                end: value.span().end,
            };
            props.push(ObjectProp { name, value, span });
            if !self.eat(TokenType::Punct, ",") {
                break;
            }
        }
        let close = self.expect_punct("}")?;
        Ok(ObjectExpr {
            props,
            lets,
            span: Span {
                start: open.start,
                end: close.end,
            },
        })
    }

    fn parse_expr(&mut self) -> PResult<Expr> {
        self.parse_ternary()
    }

    fn parse_ternary(&mut self) -> PResult<Expr> {
        let cond = self.parse_binary_level(0)?;
        if !self.eat(TokenType::Punct, "?") {
            return Ok(cond);
        }
        let then = self.parse_expr()?;
        self.expect_punct(":")?;
        let els = self.parse_expr()?;
        let span = Span {
            start: cond.span().start,
            end: els.span().end,
        };
        Ok(Expr::Cond {
            cond: Box::new(cond),
            then: Box::new(then),
            els: Box::new(els),
            span,
        })
    }

    /// Binary levels, loosest to tightest — mirrors the TS parser's chain
    /// parseNullish → … → parseMultiplicative.
    fn parse_binary_level(&mut self, level: usize) -> PResult<Expr> {
        const LEVELS: &[&[&str]] = &[
            &["??"],
            &["||"],
            &["&&"],
            &["==", "!="],
            &["<", "<=", ">", ">="],
            &["+", "-"],
            &["*", "/", "%"],
        ];
        if level >= LEVELS.len() {
            return self.parse_unary();
        }
        let ops = LEVELS[level];
        let mut left = self.parse_binary_level(level + 1)?;
        while self.peek().type_ == TokenType::Punct && ops.contains(&self.peek().value.as_str()) {
            let op = self.next().value;
            let right = self.parse_binary_level(level + 1)?;
            let span = Span {
                start: left.span().start,
                end: right.span().end,
            };
            left = Expr::Binary {
                op,
                left: Box::new(left),
                right: Box::new(right),
                span,
            };
        }
        Ok(left)
    }

    fn parse_unary(&mut self) -> PResult<Expr> {
        let t = self.peek().clone();
        if self.at(TokenType::Punct, "!") || self.at(TokenType::Punct, "-") {
            self.next();
            let operand = self.parse_unary()?;
            let span = Span {
                start: t.start,
                end: operand.span().end,
            };
            return Ok(Expr::Unary {
                op: t.value.chars().next().unwrap(),
                operand: Box::new(operand),
                span,
            });
        }
        self.parse_postfix()
    }

    fn parse_postfix(&mut self) -> PResult<Expr> {
        let mut expr = self.parse_primary()?;
        loop {
            if self.at(TokenType::Punct, ".") || self.at(TokenType::Punct, "?.") {
                let optional = self.peek().value == "?.";
                self.next();
                let name_tok = self.expect_ident()?;
                let span = Span {
                    start: expr.span().start,
                    end: name_tok.end,
                };
                expr = Expr::Member {
                    object: Box::new(expr),
                    name: name_tok.value,
                    optional,
                    span,
                };
            } else if self.at(TokenType::Punct, "[") {
                self.next();
                let inner = self.parse_expr()?;
                let close = self.expect_punct("]")?;
                let span = Span {
                    start: expr.span().start,
                    end: close.end,
                };
                expr = Expr::Bracket {
                    object: Box::new(expr),
                    inner: Box::new(inner),
                    span,
                };
            } else if self.at(TokenType::Punct, "^") {
                self.next();
                self.expect_punct("(")?;
                let mut terms: Vec<SortTerm> = Vec::new();
                while !self.at(TokenType::Punct, ")") {
                    let mut descending = false;
                    if self.at(TokenType::Punct, ">") {
                        self.next();
                        descending = true;
                    } else if self.at(TokenType::Punct, "<") {
                        self.next();
                    }
                    terms.push(SortTerm {
                        key: self.parse_expr()?,
                        descending,
                    });
                    if !self.eat(TokenType::Punct, ",") {
                        break;
                    }
                }
                let close = self.expect_punct(")")?;
                if terms.is_empty() {
                    return Err(self.fail("A sort '^(...)' needs at least one key."));
                }
                let span = Span {
                    start: expr.span().start,
                    end: close.end,
                };
                expr = Expr::Sort {
                    object: Box::new(expr),
                    terms,
                    span,
                };
            } else if self.at(TokenType::Punct, "->") {
                self.next();
                // Optional element alias / index binder, only when followed by
                // the projection body.
                let mut binder: Option<String> = None;
                let mut index_binder: Option<String> = None;
                if self.peek().type_ == TokenType::Ident
                    && self.peek_at(1).type_ == TokenType::Punct
                    && (self.peek_at(1).value == "{" || self.peek_at(1).value == ",")
                {
                    binder = Some(self.next().value);
                    if self.eat(TokenType::Punct, ",") {
                        index_binder = Some(self.expect_ident()?.value);
                    }
                }
                let body = self.parse_object_expr()?;
                let span = Span {
                    start: expr.span().start,
                    end: body.span.end,
                };
                expr = Expr::Project {
                    object: Box::new(expr),
                    binder,
                    index_binder,
                    body,
                    span,
                };
            } else {
                return Ok(expr);
            }
        }
    }

    fn parse_primary(&mut self) -> PResult<Expr> {
        let t = self.peek().clone();

        if t.type_ == TokenType::Number {
            self.next();
            return Ok(Expr::Lit {
                value: LitValue::Num(t.value.parse::<f64>().unwrap_or(f64::NAN)),
                span: Span {
                    start: t.start,
                    end: t.end,
                },
            });
        }
        if t.type_ == TokenType::Str {
            self.next();
            return Ok(Expr::Lit {
                value: LitValue::Str(t.value),
                span: Span {
                    start: t.start,
                    end: t.end,
                },
            });
        }
        if t.type_ == TokenType::Ident {
            if t.value == "true" || t.value == "false" {
                self.next();
                return Ok(Expr::Lit {
                    value: LitValue::Bool(t.value == "true"),
                    span: Span {
                        start: t.start,
                        end: t.end,
                    },
                });
            }
            if t.value == "null" {
                self.next();
                return Ok(Expr::Lit {
                    value: LitValue::Null,
                    span: Span {
                        start: t.start,
                        end: t.end,
                    },
                });
            }
            self.next();
            // Function call: `name(args)`
            if self.at(TokenType::Punct, "(") {
                self.next();
                let mut args: Vec<Expr> = Vec::new();
                while !self.at(TokenType::Punct, ")") {
                    args.push(self.parse_expr()?);
                    if !self.eat(TokenType::Punct, ",") {
                        break;
                    }
                }
                let close = self.expect_punct(")")?;
                return Ok(Expr::Call {
                    name: t.value,
                    args,
                    span: Span {
                        start: t.start,
                        end: close.end,
                    },
                });
            }
            return Ok(Expr::Ident {
                name: t.value,
                span: Span {
                    start: t.start,
                    end: t.end,
                },
            });
        }
        if self.at(TokenType::Punct, "(") {
            self.next();
            let mut inner = self.parse_expr()?;
            let close = self.expect_punct(")")?;
            inner.set_span(Span {
                start: t.start,
                end: close.end,
            });
            return Ok(inner);
        }
        if self.at(TokenType::Punct, "{") {
            return Ok(Expr::Object(self.parse_object_expr()?));
        }
        if self.at(TokenType::Punct, "[") {
            self.next();
            let mut elements: Vec<Expr> = Vec::new();
            while !self.at(TokenType::Punct, "]") {
                elements.push(self.parse_expr()?);
                if !self.eat(TokenType::Punct, ",") {
                    break;
                }
            }
            let close = self.expect_punct("]")?;
            return Ok(Expr::Array {
                elements,
                span: Span {
                    start: t.start,
                    end: close.end,
                },
            });
        }
        Err(self.fail("Expected an expression."))
    }
}
